### ISSUE TITLE
Need to get my JSON response when API fails.

### ISSUE THREAD

**Issue by hakanyildizay** [aws-amplify/amplify-ios#655]
**State your question**
I try to get response data when API fails (http status code : 4xx, 5xx). But I couldn't reach the data. The only thing thing that I can see is HttpURLResponse which contains only header values. 
 
**Which AWS Services are you utilizing?**
I am using Amplify (APIGateway)

**Provide code snippets (if applicable)**

```swift
func fetch<T:Decodable,E:Encodable>(method:HTTPMethod, service:String, input:E, resultBlock : @escaping (Result<T,HTTPError>)->()) {
    
    var headerParameters = [
        "Content-Type": "application/json",
        "Accept": "application/json",
    ]
    
    var userModelDict = input.toDictionary()
    var queryParameters : [String:String]? = nil
    var httpBodyValue: Data? = nil
        
    let responseCallBack: (AmplifyOperation<RESTOperationRequest, Data,  APIError>.OperationResult)->(Void) = { (result) in
        
        do {
            let data = try result.get()   //When http fails. This part throws an exception..
            let responseString = String(bytes: data, encoding: .utf8) ?? ""
            print(responseString)
        } catch (let e){
            print("Unable to get data \(e)")  
        }
        
        
        switch result {
            
        case .success(let data):
            self.handleResponse(response: data, completionBlock: resultBlock)
        case .failure(let apiError):
            
            switch apiError {
                
            case .networkError(let errorDesc, let userInfo, let error):
                log.debug(errorDesc)
                log.debug(userInfo)
                log.debug(error)
                break;
                
            case .httpStatusError(let statusCode, let httpUrlResponse):
                log.debug(httpUrlResponse)    //->This line is executed and HttpURLResponse does not containd any JSON Response Data.
                log.debug(statusCode)       //I can confirm that the confirmation code is 4xx or 5xx 
                break;
                
            default:
                break;
            }
            
            resultBlock(.failure(HTTPError.businessError(apiError.errorDescription)))
        }
    }
    
    let request = RESTRequest(path: service,
                              headers: headerParameters,
                              queryParameters: queryParameters,
                              body: httpBodyValue)
    
    
    switch method {
    case .get:
        _ = Amplify.API.get(request: request, listener: responseCallBack)
    case .post:
        _ = Amplify.API.post(request: request, listener: responseCallBack)
    case .put:
        _ = Amplify.API.put(request: request, listener: responseCallBack)
    default:
        log.debug("Unhandled method")
        break;
    }
    
}
```

**Environment(please complete the following information):**
 - SDK Version: Amplify (1.0.4), AmplifyPlugins/AWSAPIPlugin (1.0.4)
 - Dependency Manager: Cocoapods 
 - Swift Version : 5.0

**Device Information (please complete the following information):**
 - Device: It doesn't matter. device or simulator.
 - iOS Version: It doesn't matter. 
 - Specific to simulators:

I have seen a very similar question for React Native. ([here](https://github.com/aws-amplify/amplify-js/issues/4144))
After getting result from http call, I tried to get response data in **_httpStatusError_** case. But it only returns http status code and httpUrlResponse. And inside the httpUrlResponse there is only http headers. How can I get http response data in case of http client error. 

Looking forward to see your comments..

Take care.

**Comment by DongQuanRui** [comment]
Hi @hakanyildizay, thanks for reporting this issue. Can you let us know how you provisioned your API Gateway and related resources? It sounds like the API plugin may be dropping the response data. We will have to investigate to see if this is the behavior. Did you verify that the non-2xx response you are getting from API Gateway does in fact contain a response body? Have you also tried to reach your API Gateway endpoint using `curl` or similar tools and see if you can get the response data back?

**Comment by hakanyildizay** [comment]
Hi @DongQuanRui 

I couldn't get what you mean about provisioning/resources of API Gateway. Can you give us some example or link?
But I can provide you a Postman req/res example. I can see that I could get response data in case of 4xx error.

<img width="833" alt="Screen Shot 2020-07-21 at 09 39 02" src="https://user-images.githubusercontent.com/1074740/88021690-a5e2d400-cb36-11ea-90ca-d7fc67c00abd.png">

**Comment by hakanyildizay** [comment]
I found the solution by debugging AWSAPIPlugin. 
Unfortunately sdk does not return the data when there is an http error. 
I had to update AWSRestOperation class. 
I need to add Data to AWSOperationResponse. 

```
struct APIOperationResponse {
    let urlError: URLError?
    let httpURLResponse: HTTPURLResponse?
    let data: Data?
    
    public init(error: Error?, response: URLResponse?, responseData: Data? = nil) {
        self.urlError = error as? URLError
        self.httpURLResponse = response as? HTTPURLResponse
        self.data = responseData
    }
}

extension APIOperationResponse {

    /// Validate the response from the service and throws APIError if invalid.
    func validate() throws {
        switch (urlError, httpURLResponse) {
        case (nil, nil):
            break
        case (.some(let error), .none):
            throw APIError.networkError(error.localizedDescription, nil, error)
        case (.none, .some(let response)):
            let statusCode = response.statusCode

            let successStatusCodes = 200 ..< 300
            if !successStatusCodes.contains(statusCode) {
                throw APIError.httpStatusError(statusCode, response, data) //Passing the incoming data in httpError case.
            }
        case (.some(let error), .some(let response)):
            let userInfo = ["HTTPURLResponse": response]
            throw APIError.networkError(error.localizedDescription, userInfo, error)
        }
    }
}
```

APIOperationResponse struct should have a Data property.
```
extension AWSRESTOperation: APIOperation {
   
    func updateProgress(_ data: Data, response: URLResponse?) {
        if isCancelled || isFinished {
            finish()
            return
        }

        let apiOperationResponse = APIOperationResponse(error: nil, response: response, responseData: data)
        do {
            try apiOperationResponse.validate()
        } catch let error as APIError {
            dispatch(result: .failure(error))
            finish()
            return
        } catch {
            dispatch(result: .failure(APIError.unknown("", "", error)))
            finish()
            return
        }

        self.data.append(data)
    }
    
}

```

APIError class's httpStatusError enum should have a Optional Data type.
```
/// Errors specific to the API Category
public enum APIError {
   
    /// A non 2xx response from the service.
    case httpStatusError(StatusCode, HTTPURLResponse, Data?)

}
```

Here is the updated implementation callback function.. 
```
let responseCallBack: (AmplifyOperation<RESTOperationRequest, Data,  APIError>.OperationResult)->(Void) = { (result) in
            
            switch result {
                
            case .success(let data):
                self.handleResponse(response: data, completionBlock: resultBlock)
            case .failure(let apiError):
                
                switch apiError {
//Getting responseData (Data) in httpStatusError enum
                case .httpStatusError(let statusCode, let response, let data):
                    log.debug(statusCode)
                    log.debug(response)
                    log.debug(data) //In here getting whole response data..
                    
                default:
                    break;
                }
                
                resultBlock(.failure(HTTPError.businessError(apiError.errorDescription)))
            }
        }
```

Btw If you want I can contribute to the repo. 
What do you think?

**Comment by DongQuanRui** [comment]
Hi, @hakanyildizay. Thanks so much for your effort. We are happy to see our customers debugging into our library and provide feedback. I probably will let my teammates take a look at your PR later

**Comment by palpatim** [comment]
It does seem like this is a good feature, since some server errors are accompanied by a meaningful payload that an app could use in troubleshooting, debugging, or recovery. It's not obvious that this is the right solution.

Adding a `Data` associated value to the `APIError.httpStatusError` error enum case will be a breaking change, as apps that were previously switching on that enum case and binding the associated values will now have to deal with the new value. E.g., this prior code would break with the proposed addition:

```swift
switch apiError {
// Have an explicit error case for each previous error case
case .httpStatusError(let statusCode, let response):
    print("error \(statusCode); response: \(response)")
// No default case
}
```

We'll need to figure out a solution that breaks neither code nor behavior. In other words, it wouldn't be acceptable to start returning an entirely new enum value in cases where we encountered a service error, since existing apps may be relying on service errors to be delivered on `httpStatusError`.

**Comment by Dellybro** [comment]
Maybe subclass NSHTTPUrlResponse, which returns data, this would help with backwards compatibility.

**Comment by Dellybro** [comment]
Paste from discord

**Example of why we need this feature:**

Well for example, we use status codes in our app based on the actual HTTP meaning, so a 404 would be for invalid route, a 401 would be unauthorized etc etc. so 400 is a "bad request" a bad request for us is kind of all encompassing when it comes to what happened in the actual route method.
So a simple use case would be what message to show the user. Since i am unable to obtain the message from the API Gateway method GET, POST, PATCH, i can't let my user know more about what happened, such as "invalid credit card", "invalid CVC" etc
So if a we query for a resource and get a 400, we don't know if the resource, doesn't exist, or there was an error. Thus once again we can not correctly act upon what happened. Since a 404 is 'page not found', and we are not sending a '404' for database resource not found

**An easy solution for backwards compatibility**
Also a simple solution could possibly be to just subclass HTTPUrlResponse for AWSHTTPUrlResponse, throw the data in there, and any other information in the future that would need to be added to the UrlResponse could be backwards compatible

**Comment by tnraddatz** [comment]
If you're using API Gateway and are running into this issue, make sure you update the API Gateway's "Gateway Responses" Default 5XX and Default 4XX header's to contain response headers -- or else Access-Control-Allow-Origin wont be sent on error responses.  

For example, set response headers for Default 5XX and 4XX to:

Response header: Access-Control-Allow-Origin 
Value: '*'

**Comment by mregnauld** [comment]
Any update on this?

**Comment by jamesonwilliams** [comment]
Related: https://stackoverflow.com/q/66162851/695787

**Comment by lawmicha** [comment]
The PR https://github.com/aws-amplify/amplify-ios/pull/1076 has been released in [1.7.0](https://github.com/aws-amplify/amplify-ios/releases/tag/v1.7.0) allows the following code snippet to retrieve the data from the response:

```swift
if case let .httpStatusError(statusCode, response) = error, let awsResponse = response as? AWSHTTPURLResponse {
    if let responseBody = awsResponse.body {
        print("Response contains a response body")
    }
}
```
